File: src/meshy/cache.mjs
// cache.mjs — single source of truth for the dream3d Meshy disk cache.
//
// Pure cache primitives shared by the best-of-N CLI (scripts/meshy-generate.mjs)
// and the pipeline asset provider (src/pipeline/meshyAssetProvider.ts). Plain ESM
// (NO TypeScript syntax) so the CLI keeps running with no build step; the matching
// hand-written type declarations live alongside in cache.d.ts.
//
// Every function here is ZERO-network and has ZERO API-key dependency: a pure
// cache hit needs neither. They only touch the local filesystem under <cacheDir>.

import { readFile, writeFile, mkdir, rm } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import { createHash } from "node:crypto";

export const DEFAULT_CACHE_DIR = join(homedir(), ".cache", "dream3d", "meshy");

// Self-test anchor: this key MUST line up with the seeded cache on disk.
export const CHECKPOINT_PROMPT = "a small wooden stool";
export const CHECKPOINT_KEY = "41b60876785e9b0c";

// --- key / prompt normalization -------------------------------------------

export function normalizePrompt(prompt) {
  return prompt.trim().toLowerCase().replace(/\s+/g, " ");
}

export function deriveKey(prompt, mode) {
  const normalized = normalizePrompt(prompt);
  return createHash("sha256").update(`${normalized}::${mode}`).digest("hex").slice(0, 16);
}

// Guards against the key scheme silently drifting away from the seeded cache.
export function assertKeySchemeIsStable() {
  const got = deriveKey(CHECKPOINT_PROMPT, "preview");
  if (got !== CHECKPOINT_KEY) {
    throw new Error(
      `Cache key scheme drifted: key("${CHECKPOINT_PROMPT}","preview")=${got}, expected ${CHECKPOINT_KEY}. ` +
        `Newly generated keys would not line up with the seeded cache.`,
    );
  }
}

// --- serialization ---------------------------------------------------------

// Byte-compatible with the seeded cache: 2-space JSON, NO trailing newline.
export function serializeCache(value) {
  return JSON.stringify(value, null, 2);
}

// --- index I/O -------------------------------------------------------------

function indexPath(cacheDir) {
  return join(cacheDir, "index.json");
}

export async function readIndex(cacheDir) {
  let raw;
  try {
    raw = await readFile(indexPath(cacheDir), "utf8");
  } catch (error) {
    if (error.code === "ENOENT") {
      return {}; // no cache yet — a legitimate empty state, not a swallowed error
    }
    throw error;
  }
  return JSON.parse(raw); // a corrupt index is a loud failure
}

export async function writeIndex(cacheDir, index) {
  await mkdir(cacheDir, { recursive: true });
  await writeFile(indexPath(cacheDir), serializeCache(index));
}

// --- candidates ------------------------------------------------------------

// Keep only candidates whose .glb is still present on disk.
export function validCandidatesOnDisk(candidates) {
  return candidates.filter((candidate) => typeof candidate.glb === "string" && existsSync(candidate.glb));
}

// The chosen candidate for an entry: the one named by entry.winner if set, else
// the first candidate. Fails loudly if a named winner is missing from the pool.
export function selectCandidate(entry) {
  if (entry.winner) {
    const winner = entry.candidates.find((candidate) => candidate.taskId === entry.winner);
    if (!winner) {
      throw new Error(`cache entry ${entry.key} names winner ${entry.winner}, but no candidate has that taskId`);
    }
    return winner;
  }
  return entry.candidates[0];
}

// --- directory marker ------------------------------------------------------

// Human-readable marker so a person can identify a hash-named dir without
// recomputing sha256. Writes <cacheDir>/<key>/meta.json = { key, prompt,
// normalizedPrompt, mode } only when it does not already exist. Idempotent, no
// network.
export async function ensureDirMeta(cacheDir, key, { prompt, normalizedPrompt, mode }) {
  const dir = join(cacheDir, key);
  await mkdir(dir, { recursive: true });
  const metaPath = join(dir, "meta.json");
  if (existsSync(metaPath)) {
    return;
  }
  await writeFile(metaPath, serializeCache({ key, prompt, normalizedPrompt, mode }));
}

// --- rebuild ---------------------------------------------------------------

// Wipes one cache entry: recursively deletes <cacheDir>/<key>/ and drops
// index[key]. Returns the updated index so the caller can keep using it without
// re-reading. No network.
export async function rebuildEntry(cacheDir, key) {
  await rm(join(cacheDir, key), { recursive: true, force: true });
  const index = await readIndex(cacheDir);
  delete index[key];
  await writeIndex(cacheDir, index);
  return index;
}

File: backend/src/meshy/cache.py
"""Single source of truth for the dream3d Meshy disk cache.

Mirrors src/meshy/cache.mjs. Every function here is zero-network and has zero
API-key dependency. Uses the exact same key scheme and file layout as the TS
implementation so existing ~/.cache/dream3d data remains compatible.
"""

import hashlib
import json
import os
from pathlib import Path
from typing import Any

DEFAULT_CACHE_DIR = Path.home() / ".cache" / "dream3d" / "meshy"

# Drift-guard anchor for the key FORMULA (not the live params).
CHECKPOINT_PROMPT = "a small wooden stool"
CHECKPOINT_PARAM_SIG = (
    '{"ai_model":"meshy-6","should_remesh":true,"target_polycount":300000,"topology":"triangle"}'
)
CHECKPOINT_KEY = "80b6483c5f285dba"


def normalize_prompt(prompt: str) -> str:
    """Normalize a prompt for stable cache keys."""
    return " ".join(prompt.strip().lower().split())


def derive_key(prompt: str, mode: str, param_sig: str) -> str:
    """Return sha256(normalizedPrompt + '::' + mode + '::' + paramSig), first 16 hex chars."""
    if not isinstance(param_sig, str) or len(param_sig) == 0:
        raise ValueError(
            f"derive_key requires a non-empty param_sig string (got {param_sig!r})"
        )
    normalized = normalize_prompt(prompt)
    digest = hashlib.sha256(
        f"{normalized}::{mode}::{param_sig}".encode("utf-8")
    ).hexdigest()
    return digest[:16]


def assert_key_scheme_is_stable() -> None:
    """Guard against silent drift in the key derivation formula."""
    got = derive_key(CHECKPOINT_PROMPT, "preview", CHECKPOINT_PARAM_SIG)
    if got != CHECKPOINT_KEY:
        raise RuntimeError(
            f"Cache key scheme drifted: key({CHECKPOINT_PROMPT!r}, 'preview', "
            f"<frozen sig>)={got}, expected {CHECKPOINT_KEY}. The key derivation "
            "formula changed; newly generated keys would not line up with "
            "previously cached entries."
        )


def serialize_cache(value: Any) -> str:
    """Serialize a value to 2-space JSON with no trailing newline."""
    return json.dumps(value, indent=2)


def _index_path(cache_dir: Path) -> Path:
    return cache_dir / "index.json"


def read_index(cache_dir: Path) -> dict[str, Any]:
    """Read the cache index, returning an empty dict if it does not exist."""
    path = _index_path(cache_dir)
    try:
        raw = path.read_text(encoding="utf-8")
    except FileNotFoundError:
        return {}
    return json.loads(raw)


def write_index(cache_dir: Path, index: dict[str, Any]) -> None:
    """Write the cache index to disk."""
    cache_dir.mkdir(parents=True, exist_ok=True)
    _index_path(cache_dir).write_text(serialize_cache(index), encoding="utf-8")


def valid_candidates_on_disk(candidates: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Keep only candidates whose .glb file is still present on disk."""
    return [
        candidate
        for candidate in candidates
        if isinstance(candidate.get("glb"), str) and Path(candidate["glb"]).exists()
    ]


def select_candidate(entry: dict[str, Any]) -> dict[str, Any]:
    """Return the chosen candidate for an entry."""
    winner_id = entry.get("winner")
    if winner_id:
        for candidate in entry["candidates"]:
            if candidate.get("taskId") == winner_id:
                return candidate
        raise RuntimeError(
            f"cache entry {entry['key']} names winner {winner_id}, "
            "but no candidate has that taskId"
        )
    return entry["candidates"][0]


def ensure_dir_meta(
    cache_dir: Path,
    key: str,
    *,
    prompt: str,
    normalized_prompt: str,
    mode: str,
) -> None:
    """Write a human-readable meta.json marker into a cache key directory."""
    dir_path = cache_dir / key
    dir_path.mkdir(parents=True, exist_ok=True)
    meta_path = dir_path / "meta.json"
    if meta_path.exists():
        return
    meta_path.write_text(
        serialize_cache(
            {"key": key, "prompt": prompt, "normalizedPrompt": normalized_prompt, "mode": mode}
        ),
        encoding="utf-8",
    )


def rebuild_entry(cache_dir: Path, key: str) -> dict[str, Any]:
    """Wipe one cache entry and return the updated index."""
    import shutil

    shutil.rmtree(cache_dir / key, ignore_errors=True)
    index = read_index(cache_dir)
    index.pop(key, None)
    write_index(cache_dir, index)
    return index
